illumos-utils/src/contract.rs

```diff
@@ -0,0 +1,387 @@
+    }
+}
+
+impl Control {
+    /// Construct a new Control for the specified contract
+    pub fn new(
+        typ: ContractType,
+        ctid: ctid_t,
+    ) -> Result<Self, crate::ExecutionError> {
+        let path = path(typ, Some(ctid), "ctl");
+        match unsafe { libc::open(path.as_ptr(), libc::O_WRONLY) } {
+            fd if fd >= 0 => Ok(Control { ctid, fd }),
+            _ => Err(err(format!(
+                "opening control path {}",
+                path.into_string().unwrap()
+            ))),
+        }
+    }
+
+    /// Acknowledge an event on the contract
+    pub fn ack(
+        &self,
+        event_id: ct_evid_t,
+    ) -> Result<(), crate::ExecutionError> {
+        match unsafe { ct_ctl_ack(self.fd, event_id) } {
+            0 => Ok(()),
+            _ => Err(err(format!("failed to acknowledge event {}", event_id))),
+        }
+    }
+
+    /// Abandon the contract
+    pub fn abandon(self) -> Result<(), crate::ExecutionError> {
+        match unsafe { ct_ctl_abandon(self.fd) } {
+            0 => Ok(()),
+            _ => Err(err(format!("failed to abandon contract {}", self.ctid))),
+        }
+    }
+}
+
+// This thread watches for critical events coming from all process
+// contracts held by sled-agent, and reaps (abandons) contracts which
+// become empty. Process contracts are used in conjunction with
+// zone_enter() in order to run commands within non-global zones, and
+// the contracts used for this come from templates that define becoming
+// empty as a critical event.
+pub fn process_contract_reaper(log: Logger) {
+    let watcher = Watcher::new(ContractType::Process);
+    loop {
+        let event = watcher.watch(&log);
+        if event.typ != CT_PR_EV_EMPTY {
+            continue;
+        }
+
+        let ctl = match Control::new(ContractType::Process, event.ctid) {
+            Ok(c) => c,
+            Err(e) => {
+                error!(&log, "Failed to open contract control: {e:?}");
+                continue;
+            }
+        };
+
+        if let Err(e) = ctl.abandon() {
+            error!(log, "{e:?}");
+        } else {
+            debug!(&log, "Abandoned contract {}", event.ctid)
+        }
+    }
+}
+
+// A Rust wrapper around the process contract template.
+#[derive(Debug)]
+pub struct Template {
+    fd: c_int,
+}
+
+impl Drop for Template {
+    fn drop(&mut self) {
+        self.clear();
+        // Ignore any error, since printing may interfere with `slog`'s
+        // structured output.
+        unsafe { libc::close(self.fd) };
+    }
+}
+
+// This follows the symlink from /dev/tfpkt0 to the full path in the /devices
+// tree.  If found, it returns the path as both a string (for error reporting)
+// and as a vector of bytes (for passing to subsequent system calls).  This
+// redundancy comes from illumos/x86_64 expecting characters to be signed,
+// while Rust expects them to be unsigned.
+fn get_tfpkt_device_path() -> Option<(String, Vec<i8>)> {
```

Comment:
`nix` might be a big help here. Also, I think we can use `PathBuf` in lieu of the duel, `(String, Vec<i8>)` pair. This could then be:

```rust
fn get_tfpkt_device_path() -> Option<PathBuf> {
    let link = nix::fcntl::readlink("/dev/tfpkt0").ok()?;
    let path = PathBuf::from(link);
    if path.starts_with("../devices/") {
        path = path.strip_prefix("../devices").expect("path starts with `../devices/`");
    }
    Some(path)
}
```

Reply:
The problem is that the call expects an array of `c_char`, which is an alias for `i8` on `illumos`, and this returns an array of `u8`.  I could do a `transmute` as well, but given the small size of the array and the infrequency with which it's called, I opted for the more vanilla approach.  I suppose I could just redefine the FFI interface to be an array of `u8` and call it a day.  It's all just the same bytestream in the end.

Reply:
Aha.  Much better.  `OsStrExt` FTW.  Thanks.